### PR TITLE
Allow private trades on market

### PR DESCRIPTION
**Pull request by kcper**
### All Submissions

* [ ] Can you post a screenshot of your changes (if applicable)?
Can't post vidoes with size > 10MB now...
https://discord.com/channels/833038505626173481/851805599628394507/862993561180307456
https://discord.com/channels/833038505626173481/851805599628394507/862993699557736448

Also added this in 3rd commit (not on videos):
image

### New Feature Submissions

* [ ] Does this relate to an existing issue, or has it been talked about prior to being done (if a major change)?
https://github.com/CryptoBlades/cryptoblades/issues/190
### Notes

Please note, your code must pass all tests and lint checks before it can be merged.

### PR Description

- added target buyer input in listing setup (marked as optional for non-private sale)
- added change target buyer button in "Search my weapons/characters" view
- if someone finds a listing on chain and tries to buy it, the transaction will not go through, because of this modifier:
```
modifier userAllowedToPurchase(IERC721 _tokenAddress, uint256 id) {
        require(listings[address(_tokenAddress)][id].targetBuyer == address(0) || listings[address(_tokenAddress)][id].targetBuyer == msg.sender, "Not target buyer");
        _;
    }
```

### CODE REVIEW

**Comment by kcper** [comment]
Not sure why the build fails, any help appreciated.

**Comment by seiyria** [comment]
UI-wise, I think this looks great.

File: frontend/src/views/Market.vue
Comment (seiyria) [inline frontend/src/views/Market.vue:612]: you could use template strings for this (`) to make it cleaner, like you did above

Comment (kcper) [reply frontend/src/views/Market.vue:612]: Sure

**Comment by poshdan** [comment]
I noticed this changes the market listing structs in the market contract.
I'm fairly certain that kind of change is not upgrade-compatible, but @raymond-h might know better.

In hindsight, we could've had some kind of identifier in there to begin with, to make it more extensible. I'm not sure what to suggest as an alternative. One possible solution is to have extension mappings for listings but it may be tricky to pack an address/nftID pair into a key. Feel free to tag me with suggestions on how to get around it, and I'll take a look at it.

**Comment by raymond-h** [comment]
@poshdan Solidity supports multi-level mappings, which is even used for tracking listings themselves in the first place. And even if it didn't, it is generally considered acceptable to merge arbitrary values by ABI-encoding and hashing the result.

Adding a field to the existing struct is indeed an issue though, which I will mention in my review.

File: contracts/NFTMarket.sol
Comment (raymond-h) [inline contracts/NFTMarket.sol:54]: Adding this field will almost certainly cause problems wrt upgradeability of the contract. Instead, there should be a new multi-level mapping, similar to `listings` below, exclusively for the `targetBuyer` of a particular listing.

File: frontend/src/views/Market.vue
Comment (raymond-h) [inline frontend/src/views/Market.vue:597]: Instead of doing `this as any`, the definition in `shims-vue.d.ts` should in theory make sure `$dialog` exists as a field already on `this` (the type of `this.$dialog` as per said file *is* just `any` anyway, but that's another issue). Would it be possible to remove the `as any` cast here?

File: frontend/src/views/Market.vue
Comment (raymond-h) [inline frontend/src/views/Market.vue:716]: I assume this cast was added because the type for `filterOutTargetBuyers` was somehow not picked up. Did you try explicitly specifying the return type on the `filterOutTargetBuyers` method itself?

**Review by raymond-h — CHANGES_REQUESTED** [review]

File: frontend/src/views/Market.vue
Comment (kcper) [reply frontend/src/views/Market.vue:597]: Tried it at first, it didn't let me. I just checked again why and it seems like 
`export default Vue.extend({` instead of `export default {` causes that issue. After trying to remove `Vue.extend(` part, `this` is broken in other scenarios, e.g. `this.contracts` - property contract does not exist on `this` component. Let me know if I should investigate that more and find a fix or leave as is.

File: frontend/src/views/Market.vue
Comment (kcper) [reply frontend/src/views/Market.vue:716]: Seems to work when I do `: Promise<string>`. I will get rid of the casts.

File: contracts/NFTMarket.sol
Comment (kcper) [reply contracts/NFTMarket.sol:54]: That's a good idea, I did not think about it!

**Comment by kcper** [comment]
Build errors, because this branch was created from failing main branch, will that be a problem during merge?

**Comment by raymond-h** [comment]
Yes, please do rebase it, thank you.

File: frontend/src/views/Market.vue
Comment (raymond-h) [reply frontend/src/views/Market.vue:597]: Going to try and get this merged today, so I'll take a look at it myself instead, thanks anyway.

**Review by raymond-h — CHANGES_REQUESTED** [review]

**Comment by raymond-h** [comment]
Don't worry about actually doing the changes, I'm gonna do them myself right now anyway (plus merge in changes from `main` to resolve the conflict).

**Comment by raymond-h** [comment]
Actually on second though, merging the changes from `main` is kinda complicated because of changes to market searching (which uses an API) and I'm not terribly familiar with either those changes or this PR's frontend changes. @kcper It might be easier for you to sort out since you wrote the code, so could you take a look at doing that after all?

**Comment by kcper** [comment]
This is finished now.
@seiyria I looked at your api changes - not sure if we're on the same page.

buyerAddress is the address of a person browsing the market thus buyerAddress will be always passed to the api request (for market GET requests). I'm expecting the response to only have such nftIds that were listed without targetBuyer specified (these are being set by default to address(0) as target buyer - 0x0000000000000000000000000000000000000000) or with targetBuyer = buyerAddress.

I haven't seen any logic like this implemented so there may have been some confusion. Correct me if I'm wrong of course.

**Comment by seiyria** [comment]
> buyerAddress is the address of a person browsing the market thus buyerAddress will be always passed to the api request (for market GET requests). I'm expecting the response to only have such nftIds that were listed without targetBuyer specified (these are being set by default to address(0) as target buyer - 0x0000000000000000000000000000000000000000) or with targetBuyer = buyerAddress.

This was not my take. I had figured that you would have a new "my trades" area that only shows up if you have trades (ie, query when the market loads, then show this).

> A case when you search for your own listings (so searching by sellerAddress = you) also needs special handling - when sellerAddress = buyerAddress (you browsing the market) then don't filter out listings with not matching targetBuyer, because they are your listings.

Based on my idea above, there would never be this situation, as you'd pass them circumstantially (sellerAddress when you want to see all buy a seller, and buyerAddress when you want to see your trades).

I can adjust it as needed if that doesn't make sense, though, so let me know.

**Comment by kcper** [comment]
Search for private trades with these buttons
image

**Comment by seiyria** [comment]
UI-wise, this looks fine to me. The conflicts need to be addressed, then we can see how it goes in test if @raymond-h agrees that it looks fine.

**Comment by newb23** [comment]
Closes #190

**Review by raymond-h — APPROVED** [review]

**Comment by kcper** [comment]
Did not test it after merge as this thing still prevents adding listings:
```
if(addFee > 0) {
    skillToken.safeTransferFrom(msg.sender, address(this), usdToSkill(addFee));
}
```
Approving any amount of SKILL spend (even >1M) still fails with "transfer amount exceeds allowance".
I'd hold off with merging until it's fixed.

**Pull request merged**
